File: LeetCode/Custom Sort String.py
class Solution:
    def customSortString(self, order: str, s: str) -> str:
        counts = dict()
        for i in s:
            counts[i] = counts.get(i , 0) + 1
        res = ""
        keys = list(counts.keys())

        #ordering as wanted.
        for letter in order:
            if letter in s:
                while counts[letter] != 0:
                    res += letter
                    counts[letter] -= 1
                if counts[letter] == 0:
                    del counts[letter]
        
        #adding left letters.
        updatedKeys = list(counts.keys())
        for i in range(len(updatedKeys)):
            while counts[updatedKeys[i]] != 0:
                res += updatedKeys[i]
                counts[updatedKeys[i]] -= 1

        return res
    

a = Solution()
print(a.customSortString(order = "cba", s = "abcd" ))
print(a.customSortString(order = "bcafg", s = "abcd" ))

File: LeetCode/CustomSortString.java
import java.util.*;

public class CustomSortString{
    
    public static String customSortString(String order, String s){
        HashMap<Character, Integer> counts = new HashMap<>();
        for(char letter: s.toCharArray() ){
            if(!counts.containsKey(letter)){
                counts.put(letter, 0);
            }
            counts.put(letter, counts.get(letter) + 1);
        }
        
        String res = "";
        for(Character letter : order.toCharArray()){
            if(s.contains(String.valueOf(letter))){
                while(counts.get(letter) != 0){
                    res += letter;
                    counts.put(letter,counts.get(letter)-1);
                }
                if(counts.get(letter) == 0){
                    counts.remove(letter);
                }
            }
        }

        Set<Character> lastKeys = counts.keySet();
        List<Character> lastKeys_List = new ArrayList<>(lastKeys);
        for(int i = 0; i < lastKeys_List.size(); i++){
            while(counts.get(lastKeys_List.get(i)) != 0){
                res += String.valueOf(lastKeys_List.get(i));
                counts.put(lastKeys_List.get(i) , counts.get(lastKeys_List.get(i)) - 1);
            }
        }

        return res;
    }
    
    public static void main(String[] args){
        
        System.out.println(customSortString("cba", "abcd"));
    }
}